src/base/sparsity_pattern.C

```diff
@@ -502,139 +496,107 @@ void Build::parallel_sync ()
-        if (my_row.empty())
-        {
-          my_row.assign (their_row.begin(), their_row.end());
-        }
-        else
+      for (auto i : IntRange<std::size_t>(0, n_rows))
         {
-          my_row.insert (my_row.end(),
-                         their_row.begin(),
-                         their_row.end());
+          const auto r = received_ids[i];
+          const auto my_r = r - local_first_dof;
 
-          // We cannot use SparsityPattern::sort_row() here because it expects
-          // the [begin,middle) [middle,end) to be non-overlapping.  This is not
-          // necessarily the case here, so use std::sort()
-          std::sort (my_row.begin(), my_row.end());
+          auto & their_row = received_rows[i];
 
-          my_row.erase(std::unique (my_row.begin(), my_row.end()), my_row.end());
-        }
+          if (need_full_sparsity_pattern)
+            {
+              auto & my_row = sparsity_pattern[my_r];
 
-        // fix the number of on and off-processor nonzeros in this row
-        n_nz[my_r] = n_oz[my_r] = 0;
+              // They wouldn't have sent an empty row
+              libmesh_assert(!their_row.empty());
 
-        for (const auto & df : my_row)
-          if ((df < local_first_dof) || (df >= local_end_dof))
-            n_oz[my_r]++;
-          else
-            n_nz[my_r]++;
-      }
-      else
-      {
-        for (const auto & df : their_row)
-          if ((df < local_first_dof) || (df >= local_end_dof))
-            n_oz[my_r]++;
+              // We can end up with an empty row on a dof that touches our
+              // inactive elements but not our active ones
+              if (my_row.empty())
+              {
+                my_row.assign (their_row.begin(), their_row.end());
+              }
+              else
+              {
+                my_row.insert (my_row.end(),
+                               their_row.begin(),
+                               their_row.end());
+
+                // We cannot use SparsityPattern::sort_row() here because it expects
+                // the [begin,middle) [middle,end) to be non-overlapping.  This is not
+                // necessarily the case here, so use std::sort()
+                std::sort (my_row.begin(), my_row.end());
+
+                my_row.erase(std::unique (my_row.begin(), my_row.end()), my_row.end());
+              }
+
+              // fix the number of on and off-processor nonzeros in this row
+              n_nz[my_r] = n_oz[my_r] = 0;
+
+              for (const auto & df : my_row)
+                if ((df < local_first_dof) || (df >= local_end_dof))
+                  n_oz[my_r]++;
+                else
+                  n_nz[my_r]++;
+            }
           else
-            n_nz[my_r]++;
+            {
+              for (const auto & df : their_row)
+                if ((df < local_first_dof) || (df >= local_end_dof))
+                  n_oz[my_r]++;
+                else
+                  n_nz[my_r]++;
```

Comment:
If we do not use `need_full_sparsity_pattern`, the memory will be over allocated ? If I understand correctly, ghosting elements are counted more than once here

Reply:
This particular PR isn't changing any behavior in this respect, the code is just being reindented.  It's always operated this way when run in parallel.

But yeah, this is a questionable decision.  When I made this code run distributed, I tried to write the new MPI parallel_sync() to mimic what Ben had already done for threaded parallelism.  We avoid even temporarily allocating a full sparsity pattern, but the cost is that we end up overallocating the eventual matrix rows corresponding to boundaries between processors' subpartitions (or between threads' element ranges).

I could be argued into changing that next year, if you wanted to test the before-and-after and make sure the tradeoff is worth it.

Reply:
I did some tests back four years ago. Extra allocated memory is not small

https://github.com/idaholab/moose/issues/7901

Reply:
Thanks!  Not sure how I missed that at the time.

That's the extra allocated memory in the final matrix with need_full off.  How much extra memory was allocated in the sparsity pattern with need_full on?

Reply:
> How much extra memory was allocated in the sparsity pattern with need_full on?

There is no extra memory when need_full is on

Reply:
When need_full is on we're allocating rows that don't get freed, and we're allocating them with extra memory, not pixie dust.  Avoiding that is the entire point of the !need_full code path!

Is it possible we're failing to free those rows properly in the !need_full case?  That "row.clear();" line where a swap-with-temporary-idiom should be instead looks embarrassing and suspicious.  But it's also not a long-term problem; practically the first thing we do after `build_sparsity()` is `if (!need_full) clear_full_sparsity()`, which ought to be properly deleting everything.

Reply:
> When need_full is on we're allocating rows that don't get freed, and we're allocating them with extra memory, not > pixie dust. Avoiding that is the entire point of the !need_full code path!

What I meant was that: when need_full is on, PETSc allocates sparsity pattern accurately. When need_full is off, `n_oz` and `n_nz` are over-estimated, and then PETSc over allocates memory. The extra allocated  memory is not negligible.

Reply:
Sure, and I understood that.  But I would also like to know whether the extra allocated memory in the sparsity pattern itself, when need_full is on, is significant or not.  To compare A versus B it is not sufficient to say A is not negligible; we also have to know the size of B.

Reply:
I agreed. To have an accurate answer for your question, we should run a heap profiling for full-on and full-off.

I will do this soon

Reply:
This isn't just idle curiosity, either; it affects what the ideal long-term solution will be.  If B is negligible then let's just delete need_full_sparsity_pattern, thereby making the code simpler *and* more efficient; win-win.  If B is large but no larger than a single matrix, then let's build a full sparsity pattern (to avoid any over allocation later), but then delete it between calculating a non-over-allocated n_nz and actually allocating our first matrix.  If B is larger then we're stuck with the current code, to avoid avoidable memory spikes during sparsity build, but if it's not much larger then we might recommend that people who are using a lot of matrices with the same sparsity pattern manually enable need_full.

Reply:
Thanks!  I'm feeling guilty that I either missed your discussion (despite being pinged in a Github comment) or let it drop off my plate in 2016.  I'd been imagining overallocation levels closer to 1% than to 10%.

Have you caught any problems with this PR as-is?  I was hoping to merge this afternoon, even if it only fixes the scalability issues and not the overallocation issues.

Reply:
Could you remind me what `comandline option` I can use to turn need-full-sparsity?

Reply:
You are free to merge the PR. If we confirm there is a big issue for overallocation, we could create another PR.

Reply:
I'm afraid I can't find a command line option.  DofMap queries the matrices that are attached to it, and if any of them needs a full sparsity pattern (due to its linear algebra package) then it turns need_full on.  Oh, hell, and it's not even a public member right now, so there's no setting it manually without a libMesh rebuild.  Sorry about that.